kpb: add new state

mrajwa opened the pull request:
This adds new state KPB_STATE_INIT_DRAINING
into KPB.

Signed-off-by: Marcin Rajwa <marcin.rajwa@linux.intel.com>

dbaluta commented on src/audio/kpb.c:
```diff
@@ -715,8 +715,11 @@ static int kpb_buffer_data(struct comp_dev *dev,
+		trace_kpb_error("kpb_buffer_data() error: wrong state! (current state %d, state log %x)",
+						 kpb->state, kpb->state_log);
 		return PPL_STATUS_PATH_STOP;
+	}
```
This part doesn't have anything to do with the introduction of a new state. You can send this as a separate patch. Even separate PR if you want it merged faster.

mrajwa replied:
Okay, separate patch seems sane as I did two unrelated things in one patch.

dbaluta commented on src/audio/kpb.c:
```diff
@@ -740,8 +743,8 @@ static int kpb_buffer_data(struct comp_dev *dev,
+							 current_time - timeout, kpb->state,
+							 kpb->state_log);
 			return -ETIME;
 		}
```
Same here. This can easily be a separate patch which can be quickly merged as it is a cosmetic change.

dbaluta commented on src/include/sof/audio/kpb.h:
```diff
@@ -61,6 +61,7 @@ enum kpb_state {
 	KPB_STATE_PREPARING,
 	KPB_STATE_RUN,
 	KPB_STATE_BUFFERING,
+	KPB_STATE_INIT_DRAINING,
```
I will really want a better description of *WHY* this is needed. Also, on the long term adding a state diagram with some nice explanations in the documentation project for the KPB component would be really nice.

mrajwa commented:
@lgirdwood, @dbaluta this patch doesn't fix any issue, it just adds new state so we can have more detailed info in logs. Right now we will see a proper KPB state change like PREPARE -> RUN -> BUFFER -> **INIT_DRAINING** -> DRAINING -> ... -> RESET 

> Btw, why do we not use the standard component states here ?

Because these are private states of KPB and generic ones don't apply here.

dbaluta commented:
> @lgirdwood, @dbaluta this patch doesn't fix any issue, it just adds new state so we can have more detailed info in logs. Right now we will see a proper KPB state change like PREPARE -> RUN -> BUFFER -> **INIT_DRAINING** -> DRAINING -> ... -> RESET

@mrajwa Oh, I see. Please add this information in the commit message.

bkokoszx commented:
@mrajwa 
Do we really need to come back to ```preserved_state```? I think we should stay in ```KPB_STATE_INIT_DRAINING``` (and add support in ```kpb_copy()```) or change to ```KPB_STATE_DRAINING```, because after ```init_draining()``` we do not have to copy data to sink (selector) buffer. Can you take a look at #2343? I've added there handling ```KPB_STATE_INIT_DRAINING_STATE``` in ```kpb_copy()```.

lgirdwood commented:
@mrajwa @bkokoszx Please see my comments in #2343, it seems the same patch pushed to both PRs atm.

mrajwa commented:
> @mrajwa
> Do we really need to come back to `preserved_state`? I think we should stay in `KPB_STATE_INIT_DRAINING` (and add support in `kpb_copy()`) or change to `KPB_STATE_DRAINING`, because after `init_draining()` we do not have to copy data to sink (selector) buffer. Can you take a look at #2343? I've added there handling `KPB_STATE_INIT_DRAINING_STATE` in `kpb_copy()`.

I am against treating *KPB_STATE_INIT_DRAINING* and *KPB_STATE_DRAINING* as one state. This is unnatural and not true and will confuse people reading the code, logs etc. As for not switching back from *KPB_STATE_INIT_DRAINING* to preserved state - I think its sane to switch back as soon as init of draining is done since as I described above this state blocks all others (internal buffering being the most crucial one here)

fredoh9 commented:
SOFCI TEST

mrajwa commented:
> @mrajwa is the state diagram pending for this PR ?

![kpb_state_diagram_v1](https://user-images.githubusercontent.com/42437674/73975513-c58a9500-4926-11ea-8756-db4506adb6f6.png)

I have UML code for it also but the question is *where* should I commit it? Into sof-dosc? I see there is no parser for uml only picture source is linked.

lgirdwood commented:
@mrajwa looks good, I'm no UML parsing expert for docs, so better ask @mmaka1. Can you convert this to ASCII from UML to insert into the header (if we cant do this then a big comment is needed stating that any state changes must also be reflected in the UML diagram). i.e. I want to make it easy for anyone who changes code to alos update diagram.

lgirdwood commented on src/audio/kpb.c:
```diff
@@ -721,7 +722,9 @@ static int kpb_buffer_data(struct comp_dev *dev,
 	tracev_kpb_with_ids(dev, "kpb_buffer_data()");
 
-	if (kpb->state != KPB_STATE_RUN && kpb->state != KPB_STATE_DRAINING) {
+	if (kpb->state != KPB_STATE_RUN &&
```
Lets just state the assumption about state in a comment here.

lgirdwood commented on src/include/sof/audio/kpb.h:
```diff
@@ -61,6 +61,7 @@ enum kpb_state {
 	KPB_STATE_PREPARING,
 	KPB_STATE_RUN,
 	KPB_STATE_BUFFERING,
+	KPB_STATE_INIT_DRAINING,
```
and we can probably have a big fat comment here that any state changes must also be reflected in ASCII header diagram or UML diagram on sof-docs (with a link to said UML diagram).

mrajwa replied:
Yes very good idea, so let me first finish UML merging and once this is done I will come back here with the update.

jajanusz replied:
Please use doxygen and insert link properly, check here: https://stackoverflow.com/a/54637991

mrajwa commented:
@lgirdwood, I tried to convert it to ASCII using plantuml built in feature but it doesn't look good. So I would include it in the sof dosc in the form of plantuml code and/or actual rendered picture.

keyonjie commented:
> @lgirdwood, I tried to convert it to ASCII using plantuml built in feature but it doesn't look good. So I would include it in the sof dosc in the form of plantuml code and/or actual rendered picture.

@mrajwa you could try asciiflow here: http://asciiflow.com/

mrajwa commented:
@keyonjie, this tool seems to be aimed to draw ASCII diagrams and it might work however, I don't see any option to save it as a code and no option to generate a picture for sof-dosc. Especially the former concerns me since it will make our diagrams hard to maintain - each time you would like to add new state you would have to redraw it yourself or tinker with ASCII in your own...nightmare. Plantuml is based on their code, so each and every time you modify something it all boils down to changing specific line in code. Not to mention it generates a nice picture right away.

lgirdwood commented:
@mrajwa  ok, lets do it in sof-docs, as long as we have some comment that explains this with command line for using the tool to generate the output.

mrajwa commented:
I have added a PR into sof-docs https://github.com/thesofproject/sof-docs/pull/183 and added comments you requested in the FW. Can we merge now? I would like to merge it fast as I want to open new, bigger PR for KWD aimed to make it even more secure in corner cases.
